serde/src/private/size_hint.rs

```diff
@@ -8,9 +8,17 @@ where
 }
 
 #[cfg(any(feature = "std", feature = "alloc"))]
-#[inline]
-pub fn cautious(hint: Option<usize>) -> usize {
-    cmp::min(hint.unwrap_or(0), 4096)
+pub fn cautious<Element>(hint: Option<usize>) -> usize {
+    const MAX_PREALLOC_BYTES: usize = 1024 * 1024;
+
+    if mem::size_of::<Element>() == 0 {
```

Comment:
It would be good to cache the result of `mem::size_of::<Element>()` and reuse it at line 19 below

Reply:
This is the [const function](https://doc.rust-lang.org/std/mem/fn.size_of.html) since 1.24 so it will be computed by the compiler